### ISSUE TITLE
Auto-selection of X11 platforms fails with Nvidia drivers (470 and 515) using LXDE

### ISSUE THREAD

**Issue by frank-dspeed** [MirServer/mir#2467]
NVIDIA Cards: 970 GTX
Nvidia Driver version: 470.129.06 
x11 version: 1.20.13 (12013000)
Os Ubuntu Latest Greatest.
Desktop: Lubuntu is a light Linux distribution using the Lightweight X11 Desktop Environment (LXDE)
```
+ exec /snap/ubuntu-frame/2294/bin/run-frame /snap/ubuntu-frame/2294/usr/local/bin/frame
+ [ /snap/ubuntu-frame/2294/usr/local/bin/frame = --help ]
+ dirname /run/user/1000/snap.ubuntu-frame
+ XDG_RUNTIME_DIR=/run/user/1000
+ export XDG_RUNTIME_DIR
+ mkdir -p /run/user/1000 -m 700
+ [ -n wayland-99 ]
+ [ -e /run/user/1000/wayland-99 ]
+ exec /snap/ubuntu-frame/2294/usr/local/bin/frame
[2022-06-13 11:10:28.044906] <information> mirserver: Starting
[2022-06-13 11:10:28.045598] < - debug - > mirserver: Not trying logind: "DISPLAY" is set and X need not have claimed the VT
[2022-06-13 11:10:28.045875] < - debug - > mirserver: Not using Linux VT subsystem for session management: Failed to find the current VT
[2022-06-13 11:10:28.045894] < - debug - > mirserver: No session management supported
[2022-06-13 11:10:28.045923] <information> VT switch key handler: No VT switching support available: MinimalConsoleServices does not support VT switching
[2022-06-13 11:10:28.046011] <information> mircommon: Loading modules from: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform
[2022-06-13 11:10:28.046039] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/graphics-gbm-kms.so.20
[2022-06-13 11:10:28.046049] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/server-x11.so.20
[2022-06-13 11:10:28.046056] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/input-evdev.so.8
[2022-06-13 11:10:28.046630] <information> mirserver: Found display driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.048335] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver nvidia
[2022-06-13 11:10:28.048448] < - debug - > gbm-kms: MIR_MESA_KMS_DISABLE_MODESET_PROBE is set
[2022-06-13 11:10:28.048457] < -warning- > gbm-kms: Failed to detect whether device /dev/dri/card0 supports KMS, continuing with lower confidence
[2022-06-13 11:10:28.048480] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.048486] <information> mirserver: 	/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card0 (priority 128)
[2022-06-13 11:10:28.048493] <information> mirserver: Found display driver: mir:x11 (version 2.8.0)
[2022-06-13 11:10:28.053422] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.053433] <information> mirserver: 	System (priority 192)
[2022-06-13 11:10:28.053470] <information> mirserver: Selected display driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.053728] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver nvidia
[2022-06-13 11:10:28.053796] < - debug - > gbm-kms: MIR_MESA_KMS_DISABLE_MODESET_PROBE is set
[2022-06-13 11:10:28.053804] < -warning- > gbm-kms: Failed to detect whether device /dev/dri/card0 supports KMS, but continuing anyway
[2022-06-13 11:10:28.053818] <information> gbm-kms: Using DRM device /dev/dri/card0
[2022-06-13 11:10:28.184679] <information> mirserver: Selected display driver: mir:x11 (version 2.8.0)
[2022-06-13 11:10:28.184880] <information> mircommon: Loading modules from: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform
[2022-06-13 11:10:28.184928] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/graphics-gbm-kms.so.20
[2022-06-13 11:10:28.184942] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/server-x11.so.20
[2022-06-13 11:10:28.184952] <information> mircommon: Loading module: /snap/ubuntu-frame/2294/usr/lib/x86_64-linux-gnu/mir/server-platform/input-evdev.so.8
[2022-06-13 11:10:28.185785] <information> mirserver: Found rendering driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.186247] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver nvidia
[2022-06-13 11:10:28.186444] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/renderD128, driver nvidia
[2022-06-13 11:10:28.201041] <information> gbm-kms: Detected software renderer: llvmpipe (LLVM 12.0.0, 256 bits)
[2022-06-13 11:10:28.201524] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.201544] <information> mirserver: 	/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/renderD128 (priority 128)
[2022-06-13 11:10:28.201554] <information> mirserver: Found rendering driver: mir:x11 (version 2.8.0)
[2022-06-13 11:10:28.201561] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.201566] <information> mirserver: 	System (priority 192)
[2022-06-13 11:10:28.201605] <information> mirserver: Selected rendering driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.201637] <information> mirserver: Selected rendering driver: mir:x11 (version 2.8.0)
ERROR: /build/mir-6HuLgA/mir-2.8.0/src/platforms/common/server/kms-utils/drm_mode_resources.cpp(39): Throw in function mir::graphics::kms::DRMModeResUPtr {anonymous}::resources_for_drm_node(int)
Dynamic exception type: boost::wrapexcept<std::system_error>
std::exception::what: Couldn't get DRM resources: Operation not supported
```

## Current Related Issues
- Current State NVIDIA: https://forums.developer.nvidia.com/t/wayland-information-for-r515-beta-release/214275
- Ubuntu State: https://discourse.ubuntu.com/t/where-does-ubuntu-frame-work/23193
- MirServer State: https://github.com/MirServer/mir/issues/1544

## Workaround
> Ubuntu Frame [does not currently include support for Nvidia drivers](https://discourse.ubuntu.com/t/where-does-ubuntu-frame-work/23193)
> 
> But this trace looks like you are trying to run hosted on another desktop that supports X11. And running on X11 could work. Can you try the following?
> 
> ```
> ubuntu-frame --platform-display-libs mir:x11 --platform-rendering-libs mir:x11
> ```

**Comment by AlanGriffiths** [comment]
Ubuntu Frame [does not currently include support for Nvidia drivers](https://discourse.ubuntu.com/t/where-does-ubuntu-frame-work/23193) 

But this trace looks like you are trying to run hosted on another desktop that supports X11. And running on X11 could work. Can you try the following?

```
ubuntu-frame --platform-display-libs mir:x11 --platform-rendering-libs mir:x11
```

**Comment by AlanGriffiths** [comment]
@frank-dspeed thanks for checking that. It clarifies what the problem is (selecting the right graphics & rendering platforms). This ought to work without specifying these platforms on the command line, and we'll look into it.

I've updated the title to reflect the underlying problem.

**Comment by frank-dspeed** [comment]
Maybe it is importent so i add to the Issue description that i am using a Nvida 970 GTX maybe there are problems that are related 
I know internal that nvidia did drop support of many Graphic Adapters as they upgraded from 3xx to 4xx while 515+ readds all the adapters so there EGL Got backported

**Comment by frank-dspeed** [comment]
@AlanGriffiths Oh i think i got now why it is not so general for you i am using LXDE! I am using the Lubuntu Desktop

**Comment by AlanGriffiths** [comment]
[These are problem solving notes for future reference, not an attempt at a solution]

It looks like the confusion arises with the gbm-kms probing of the nvidia card here:
```
[2022-06-13 11:10:28.046630] <information> mirserver: Found display driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.048335] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver nvidia
[2022-06-13 11:10:28.048448] < - debug - > gbm-kms: MIR_MESA_KMS_DISABLE_MODESET_PROBE is set
[2022-06-13 11:10:28.048457] < -warning- > gbm-kms: Failed to detect whether device /dev/dri/card0 supports KMS, continuing with lower confidence
[2022-06-13 11:10:28.048480] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.048486] <information> mirserver: 	/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card0 (priority 128)
```
That rates `gbm-kms` as `PlatformPriority::supported`

By comparison, on a laptop using the i915 intel driver we see the following:
```
[2022-06-15 09:43:22.741636] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver i915
[2022-06-15 09:43:22.741959] <information> gbm-kms: Failed to probe DRM device: /build/mir-ygV1mk/mir-2.8.0+dev225-gd7826d3c16/src/server/console/minimal_console_services.cpp(134): Throw in function virtual std::future<std::unique_ptr<mir::Device> > mir::MinimalConsoleServices::acquire_device(int, int, std::unique_ptr<mir::Device::Observer>)
Dynamic exception type: boost::wrapexcept<std::system_error>
std::exception::what: Failed to acquire DRM master: Operation not permitted
```

We then go on to select the  (`PlatformPriority::hosted`) `mir:x11` display platform as expected:
```
[2022-06-13 11:10:28.184679] <information> mirserver: Selected display driver: mir:x11 (version 2.8.0)
```

But when we go on to the rendering platform we see:
```
[2022-06-13 11:10:28.185785] <information> mirserver: Found rendering driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.186247] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/card0, driver nvidia
[2022-06-13 11:10:28.186444] < - debug - > gbm-kms: Quirks: checking device with devnode: /dev/dri/renderD128, driver nvidia
[2022-06-13 11:10:28.201041] <information> gbm-kms: Detected software renderer: llvmpipe (LLVM 12.0.0, 256 bits)
[2022-06-13 11:10:28.201524] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.201544] <information> mirserver: 	/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/renderD128 (priority 128)
[2022-06-13 11:10:28.201554] <information> mirserver: Found rendering driver: mir:x11 (version 2.8.0)
[2022-06-13 11:10:28.201561] <information> mirserver: Driver supports:
[2022-06-13 11:10:28.201566] <information> mirserver: 	System (priority 192)
[2022-06-13 11:10:28.201605] <information> mirserver: Selected rendering driver: mir:gbm-kms (version 2.8.0)
[2022-06-13 11:10:28.201637] <information> mirserver: Selected rendering driver: mir:x11 (version 2.8.0)
```
We select both the (`PlatformPriority::supported`)  `gbm-kms` and the (`PlatformPriority::hosted`) `mir:x11` rendering platforms. It is the subsequent attempt to use the (lower priority) `gbm-kms` platform that causes the failure. 

I'm not sure why this rendering platform is selected along with `x11`. (Possibly something to do with hybrid support? Possibly simply incorrect?)

**Comment by AlanGriffiths** [comment]
Hmm, this is where the failure happens:
```C++
                // TODO: Do we want to be able to continue on partial failure here?
                rendering_platforms.push_back(
                    create_rendering_platform(
                        device,
                        the_display_platforms(),
                        *the_options(),
                        *the_emergency_cleanup()));
```
I think we have the answer to the "TODO" :grin:

**Comment by frank-dspeed** [comment]
I will investigate and also use my intel onboard graphics and connect a monitor to that to see what happens as soon as you tell me then we can look into the hybrid support stuff

**Comment by AlanGriffiths** [comment]
> the hybrid support stuff

The hybrid support isn't complete. (Yet.) But it does complicate the logic for selecting display and rendering platforms.

**Comment by AlanGriffiths** [comment]
I've now got my system running on Nvidia and can reproduce. In view of the above analysis I tried the effect of:
```
ubuntu-frame --platform-rendering-libs mir:x11
```
But this still fails to start.

Surprisingly, (as the display platform was already correctly selected) _this_ works:
```
ubuntu-frame --platform-display-libs mir:x11
```

**Comment by AlanGriffiths** [comment]
Update: the above analysis led me to an incorrect hypothesis.

If I build and run Frame outside the snap container it works as expected and, despite my concerns above, there are multiple rendering platforms selected. The same with various other Mir based servers.

Unpacking the snap and using the binaries from within it also works. (Here's the command I used for reference:
```
unsquashfs -d try-me ubuntu-frame_2603.snap
LD_LIBRARY_PATH=/snap/mesa-core20/current/egl/lib:try-me/usr/lib/ LIBGL_DRIVERS_PATH=/snap/mesa-core20/current/egl/dri __EGL_VENDOR_LIBRARY_DIRS=/snap/mesa-core20/current/egl/glvnd/egl_vendor.d/  MIR_SERVER_PLATFORM_PATH=try-me/usr/lib/x86_64-linux-gnu/mir/server-platform/ try-me/usr/local/bin/frame
```

Running the snap unconfined _doesn't_ work, so it probably isn't a confinement issue.

**Comment by AlanGriffiths** [comment]
Update: after getting some "works when expected to fail" surprises, I _think_ (not figured out how to arrange this on demand) that problems only occur when the `gbm-kms` driver is probed (found on the filesystem?) before the `x11` driver.

**Comment by AlanGriffiths** [comment]
There's a bunch of interacting problems here:

1. The order in which platforms are considered is significant (not yet figured out why)
2. The `gbm-kms` platform on my system thinks (wrongly) that it has `PlatformPriority::best` support for the Nvidia card (which means sorting by priority doesn't work)
3. If the `gbm-kms` is considered first then it fails later in startup (because it tries to use the Nvidia card in a way that doesn't currently work)

But all this means that I have a simpler workaround that I can add to the default snap configuration:
```
$ snap set ubuntu-frame config="driver-quirks=skip:driver:nvidia"
```

**Comment by frank-dspeed** [comment]
i would say lets try them all and return first working

**Comment by AlanGriffiths** [comment]
OK, I've discussed briefly with @RAOF who is working on the multi-platform support and the underlying issue with selecting the right "hosted" platforms will be swept up with that.

In the meantime, for Ubuntu Frame I've [incorporated a workaround](https://github.com/MirServer/ubuntu-frame/pull/74) that resolves things well enough for now.

@frank-dspeed, you can test the workaround out with:
```
snap refresh --channel=edge/pr74 ubuntu-frame
```
That will likely take around a week to reach `stable`.

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths works like a charm

@frank-dspeed, if you could comment on [the PR](https://github.com/MirServer/ubuntu-frame/pull/74) it might help speed the review process.

**Comment by frank-dspeed** [comment]
only for people that look into this because the find the issue 

this is closed when: by https://github.com/MirServer/ubuntu-frame/commit/95aa46a8689b85b60a298299e14a1e504f752aee

and is already merged expect it with the next release.

**Comment by AlanGriffiths** [comment]
Reopening as the problem still exists in Mir and there's only a partial workaround in Frame

**Comment by AlanGriffiths** [comment]
Solved by #2556
